dialect/sql/schema/migrate.go

```diff
@@ -152,496 +117,521 @@ func NewMigrate(d dialect.Driver, opts ...MigrateOption) (*Migrate, error) {
-			m.fileTypeRanges = nil
-			m.dbTypeRanges = nil
+			// We clean a database by dropping all tables inside the connected schema.
+			s, err = a.atDriver.InspectSchema(ctx, "", nil)
+			if err != nil {
+				return
+			}
+			tbls := make([]schema.Change, len(s.Tables))
+			for i, t := range s.Tables {
+				tbls[i] = &schema.DropTable{T: t}
+			}
+			if err2 := a.atDriver.ApplyChanges(ctx, tbls); err2 != nil {
+				if err != nil {
+					err = fmt.Errorf("%v: %w", err2, err)
+					return
+				}
+				err = err2
+				return
+			}
 		}()
+		// Replay the migration directory on the database.
+		ex, err := migrate.NewExecutor(a.atDriver, a.dir, &migrate.NopRevisionReadWriter{})
+		if err != nil {
+			return err
+		}
+		if err := ex.ExecuteN(ctx, 0); err != nil && !errors.Is(err, migrate.ErrNoPendingFiles) {
+			return err
+		}
+	default:
+		return fmt.Errorf("unknown migration mode: %q", a.mode)
 	}
-	m.setupTables(tables)
-	plan, err := m.atDiff(ctx, m, name, tables...)
+	plan, err := a.plan(ctx, a.sqlDialect, name, tables)
 	if err != nil {
 		return err
 	}
-	if m.universalID {
-		newTypes := m.typeRanges[len(m.dbTypeRanges):]
-		if len(newTypes) > 0 {
-			plan.Changes = append(plan.Changes, &migrate.Change{
-				Cmd:     m.atTypeRangeSQL(newTypes...),
-				Comment: fmt.Sprintf("add pk ranges for %s tables", strings.Join(newTypes, ",")),
-			})
-		}
-	}
 	// Skip if the plan has no changes.
 	if len(plan.Changes) == 0 {
 		return nil
 	}
-	return migrate.NewPlanner(nil, m.atlas.dir, opts...).WritePlan(plan)
+	return migrate.NewPlanner(nil, a.dir, opts...).WritePlan(plan)
 }
 
-func (m *Migrate) create(ctx context.Context, tables ...*Table) error {
-	tx, err := m.Tx(ctx)
-	if err != nil {
-		return err
+// WithDialect configures the Ent dialect to use when migrating for an Atlas supported dialect flavor.
+// As an example, Ent can work with TiDB in MySQL dialect and Atlas can handle TiDB migrations.
+func WithDialect(d string) MigrateOption {
+	return func(a *Atlas) {
+		a.dialect = d
 	}
-	if err := m.init(ctx, tx); err != nil {
-		return rollback(tx, err)
+}
+
+// WithGlobalUniqueID sets the universal ids options to the migration.
+// Defaults to false.
+func WithGlobalUniqueID(b bool) MigrateOption {
+	return func(a *Atlas) {
+		a.universalID = b
 	}
-	if m.universalID {
-		if err := m.types(ctx, tx); err != nil {
-			return rollback(tx, err)
-		}
+}
+
+// WithUniversalID instructs atlas to use a file based type store when
+// global unique ids are enabled. For more information see the setupAtlas method on Migrate.
+//
+// ATTENTION:
+// The file based PK range store is not backward compatible, since the allocated ranges were computed
+// dynamically when computing the diff between a deployed database and the current schema. In cases where there
+// exist multiple deployments, the allocated ranges for the same type might be different from each other,
+// depending on when the deployment took part.
+func WithUniversalID() MigrateOption {
```

Comment:
This should be deleted, no?

Reply:
We can indeed remove this and use the `GlobalUniqueID` instead (as we discussed)